controller/object_store_controller.go

```diff
@@ -0,0 +1,1139 @@
+	return nil
+}
+
+// This function does a short sanity check on the various resources that are
+// needed to operate the object stores. If any of them is found to be
+// unhealthy, the controller will transition the object store to "Error"
+// state, otherwise do nothing.
+func (osc *ObjectStoreController) handleRunning(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateStopped {
+		store.Status.State = longhorn.ObjectStoreStateStopping
+		return nil
+	}
+
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil || dpl.Status.UnavailableReplicas > 0 {
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err != nil {
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+
+	_, err = osc.ds.GetIngress(osc.namespace, store.Name)
+	if err != nil {
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+
+	pvc, err := osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
+	if err != nil || pvc.Status.Phase != corev1.ClaimBound {
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+
+	_, err = osc.ds.GetVolume(genPVName(store))
+	if err != nil {
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolume(genPVName(store))
+	if err != nil {
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopping(store *longhorn.ObjectStore) (err error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+	} else if *dpl.Spec.Replicas != 0 {
+		dpl.Spec.Replicas = int32Ptr(0)
+		_, err = osc.ds.UpdateDeployment(dpl)
+		return err
+	} else if dpl.Status.AvailableReplicas > 0 {
+		return nil // wait for shutdown
+	}
+
+	store.Status.State = longhorn.ObjectStoreStateStopped
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopped(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateRunning {
+		store.Status.State = longhorn.ObjectStoreStateStarting
+		return nil
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleTerminating(store *longhorn.ObjectStore) (err error) {
+	// Wait for all directly owmed resources to be deleted before removing the finalizer
+
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetIngress(osc.namespace, store.Name)
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetDeployment(store.Name)
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolume(genPVName(store))
```

Comment:
Note that PV has cluster scope so we cannot rely on the background deletion mechanism of Kubenetes. Maybe we need to manually delete the PV here. Ref https://github.com/longhorn/longhorn-manager/pull/2136#discussion_r1376807518

Reply:
In my testing, the PV always was automatically deleted (I assume by the CSI driver). It is explicitly bound to the Longhorn volume, which has an owner reference and is garbage collected.
Did you observe something else?

Reply:
I observed the same behavior: PV got deleted when we delete the ObjectStore. It was super confused to me as I was wondering who deleted the PV. There is no owner reference so it cannot be Kubernetes garbage collection. It is also not the longhorn-csi-plugin because the PV has `PersistentVolumeReclaimPolicy: corev1.PersistentVolumeReclaimRetain`. 

Eventually, I found that volume controller deletes it [here](https://github.com/longhorn/longhorn-manager/blob/0258c74e026a2ad65ea03f8fc9cb412597e53cdd/controller/volume_controller.go#L362-L368). 

However, volume controller does not always delete it, it only deletes PV when the `volume.Status` successfully observes the PV. In some rare cases, volume controller might miss deleting the PV. Therefore, I think we should still delete the PV inside the handleTermination() function before remove the finalizer for the ObjectStore

Reply:
I think we need to make up our minds here. If the volume controller deletes the PV, then the volume controller should _reliably_ delete the PV and the object store controller should just expect that.
Or if the volume controller does not reliably delete the PV, it should not delete the PV (for the object store) ever, leaving it to the object store controller in all cases.
I think either way is a fair approach, but we have to be deliberate and can't just let the two controllers mingle.
Otherwise this is going to be a mess with race-conditions between the controllers and/or edge cases where neither controller is able to delete the PV.

Reply:
I see your point about wanting to have the cleanup process inside a single controller. Even though I don't see a potential race condition here (each controller will try to issue a PV delete request, the delete request is idempotent so it should be fine), I agree that we should should bring under 1 controller. 

Since the volume controller cannot guarantee to always delete the PV (until it appear in the Volume.Status), can we do:
1. Exclude the ObjectStore PV deletion inside volume controller
2. Add PV deletion to ObjectStore controller inside handleTerminating before removing the finalizer

Reply:
> each controller will try to issue a PV delete request, the delete request is idempotent so it should be fine
One of the controllers would see an "is not found" error. This should of course be handled gracefully in either case, but it will probably be logged to the console.

Since we can be reasonably sure that both controllers are in the same manager process (since the controller responsible for the object store has to have the same name as the volume controller responsible for the volume), I think we can assume that both controllers see the same information in the cache. Therefore it should be possible to just check the volume's `.status.PVName` field in the object store controller remove the PV if the field is found to be empty.
If the field is not found to be empty, the volume controller would see the same (nonempty) field in the cache and delete the PV.

What do you think? Is that a valid assumption? Would it be too fragile or would it be fine?

Reply:
Instead of adding an exception for the ObjectStore PV deletion inside the volume controller, can we make the volume controller should reliably delete the PV?

Reply:
> Instead of adding an exception for the ObjectStore PV deletion inside the volume controller, can we make the volume controller should reliably delete the PV?

I think this one is a good approach. We can modify the volume controller to delete 2 things:
1. PV inside volume.status
2. PV of objectstore if this volume is a backed volume for an object store

Then, volume controller will reliably delete all potential PVs


> Since we can be reasonably sure that both controllers are in the same manager process (since the controller responsible for the object store has to have the same name as the volume controller responsible for the volume), I think we can assume that both controllers see the same information in the cache. Therefore it should be possible to just check the volume's .status.PVName field in the object store controller remove the PV if the field is found to be empty.
> If the field is not found to be empty, the volume controller would see the same (nonempty) field in the cache and delete the PV.
> 
> What do you think? Is that a valid assumption? Would it be too fragile or would it be fine?


I think this is a bit fragile as the 2 controllers might see 2 different versions of the volume objects while reconciling:
1. Volume controller pickup volume object, version v1.
2. Volume controller goroutine is put to sleep by OS
3. Cache is updated, volume object has v2 version
4. ObjectStore pickup the volume object, version v2
5. OS resume volume controller goroutine
6. Now, 2 controllers are seeing 2 different versions of the volume object

Reply:
If we go with explicit foreground cascading deletion as suggested [here](https://github.com/longhorn/longhorn-manager/pull/2136#discussion_r1401103561), the object store controller will delete the PV anyways and the volume controller doesn't need to clean it up anymore.

This seems to me like it's the best way, because we don't need to modify the volume controller.